### ISSUE TITLE
The results of the onnx model show positional deviation

### ISSUE THREAD

**Issue by w1005444804** [facebookresearch/segment-anything#130]
What is the reason for this situation to occur？
![Figure_1](https://user-images.githubusercontent.com/6340026/230782299-ee299712-42c5-41bc-9c0c-83e17678b20f.png)

**Comment by FullStackSimon** [comment]
What does your code look like?

**Comment by w1005444804** [comment]
@FullStackSimon 
import torch
import numpy as np
import cv2
import os
import matplotlib.pyplot as plt
from segment_anything.utils.transforms import ResizeLongestSide
from torch.nn import functional as F

import onnxruntime


def show_mask(mask, ax):
    color = np.array([30 / 255, 144 / 255, 255 / 255, 0.6])
    h, w = mask.shape[-2:]
    mask_image = mask.reshape(h, w, 1) * color.reshape(1, 1, -1)
    ax.imshow(mask_image)


def show_points(coords, labels, ax, marker_size=375):
    pos_points = coords[labels == 1]
    neg_points = coords[labels == 0]
    ax.scatter(pos_points[:, 0], pos_points[:, 1], color='green', marker='*', s=marker_size, edgecolor='white',
               linewidth=1.25)
    ax.scatter(neg_points[:, 0], neg_points[:, 1], color='red', marker='*', s=marker_size, edgecolor='white',
               linewidth=1.25)


def show_box(box, ax):
    x0, y0 = box[0], box[1]
    w, h = box[2] - box[0], box[3] - box[1]
    ax.add_patch(plt.Rectangle((x0, y0), w, h, edgecolor='green', facecolor=(0, 0, 0, 0), lw=2))


image = cv2.imread('notebooks/images/sux.jpg')
image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

plt.figure(figsize=(10,10))
plt.imshow(image)
plt.axis('on')
plt.show()


encoder_path = "H:/sam/encoder.onnx"
decoder_path = "H:/sam/decoder.onnx"

image_size = 1024
transform = ResizeLongestSide(image_size)

input_image = transform.apply_image(image)
input_image_torch = torch.as_tensor(input_image, device="cpu")
input_image_torch = input_image_torch.permute(2, 0, 1).contiguous()[None, :, :, :]
pixel_mean = torch.Tensor([123.675, 116.28, 103.53]).view(-1, 1, 1)
pixel_std = torch.Tensor([58.395, 57.12, 57.375]).view(-1, 1, 1)
x = (input_image_torch - pixel_mean) / pixel_std
h, w = x.shape[-2:]
padh = image_size - h
padw = image_size - w
x = F.pad(x, (0, padw, 0, padh))
x = x.numpy()

encoder_inputs = {
    "x": x,
}

image_embedding = []
image_embedding_filename = "image_embedding.npy"
if os.path.exists(image_embedding_filename):
    image_embedding = np.load('image_embedding.npy')
else:
    encoder_session = onnxruntime.InferenceSession(encoder_path)
    output = encoder_session.run(None, encoder_inputs)
    image_embedding = output[0]
    np.save('image_embedding', image_embedding)

decoder_session = onnxruntime.InferenceSession(decoder_path)
#-----------------------------------------------------------
input_point = np.array([[690, 430], [642, 256]])
input_label = np.array([1,1])


onnx_coord = np.concatenate([input_point, np.array([[0.0, 0.0]])], axis=0)[None, :, :]
onnx_label = np.concatenate([input_label, np.array([-1])], axis=0)[None, :].astype(np.float32)

onnx_coord = transform.apply_coords(onnx_coord, image.shape[:2]).astype(np.float32)

onnx_mask_input = np.zeros((1, 1, 256, 256), dtype=np.float32)
onnx_has_mask_input = np.zeros(1, dtype=np.float32)

decoder_inputs = {
    "image_embeddings": image_embedding,
    "point_coords": onnx_coord,
    "point_labels": onnx_label,
    "mask_input": onnx_mask_input,
    "has_mask_input": onnx_has_mask_input,
    "orig_im_size": np.array(image.shape[:2], dtype=np.float32)
}

masks, iou, low_res_logits = decoder_session.run(None, decoder_inputs)

masks = (masks > 0.0)
c, h, w = masks.shape[-3:]
masksx = np.split(masks, c, axis=1)
plt.figure(figsize=(10, 10))
plt.imshow(image)
#os.mkdir("save")
for i in range(c):
    show_mask(masksx[i], plt.gca())
    name = "save/m{}.png".format(i)
    mat = masksx[i].reshape(h, w, 1).astype(np.uint8) * 255
    cv2.imwrite(name, mat)

cv2.imwrite("image.jpg", image)
show_points(input_point, input_label, plt.gca())
plt.axis('off')
plt.show()

**Comment by w1005444804** [comment]
pytorch do:
![Figure_2](https://user-images.githubusercontent.com/6340026/230782796-5abc4be1-f52b-47eb-8657-2c508a2dd488.png)

**Comment by w1005444804** [comment]
Most of the code comes from：
[https://github.com/visheratin/segment-anything/blob/main/notebooks/onnx_model_example.ipynb](url)

**Comment by FullStackSimon** [comment]
Sorry I'm going to have to box out - I don't know anything about onnx

**Comment by w1005444804** [comment]
@FullStackSimon thanks

**Comment by zdhernandez** [comment]
@w1005444804  See https://github.com/facebookresearch/segment-anything/issues/115 and my code is there in the Zip (which is pretty much the ONNX example)

**Comment by EndingCredits** [comment]
@w1005444804 see #56 it's an issue in how the ONNX model is created. I have submitted a PR that should fix this if you regenerate the ONNX model.

**Comment by zdhernandez** [comment]
@w1005444804  thanks!

**Comment by w1005444804** [comment]
When I set the aspect ratio of the image to be the same as the aspect ratio of the dog image，things changed！？？？？
![Figure_3](https://user-images.githubusercontent.com/6340026/230811575-e3c6342d-e230-4dcb-a624-d250dcc7bed3.png)

**Comment by eeoa** [comment]
@w1005444804 

How you get the onnx models "encoder.onnx" and "decoder.onnx" for image embedding?

**Comment by nikhilaravi** [comment]
@w1005444804 please try re-exporting the onnx model to use the fix from https://github.com/facebookresearch/segment-anything/issues/56. 

If you still have issues please feel free to reopen the issue.

**Comment by zdhernandez** [comment]
@w1005444804  was this fixed at all ? how did you change the aspect ratio ?
